src/server/pkg/migrations/migrations.go

```diff
@@ -68,28 +68,44 @@ func InitialState() State {
 	return State{
 		name: "init",
 		change: func(ctx context.Context, env Env) error {
-			_, err := env.Tx.ExecContext(ctx, `CREATE TABLE IF NOT EXISTS migrations (
-				id BIGINT PRIMARY KEY,
-				NAME VARCHAR(250) NOT NULL,
-				start_time TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
-				end_time TIMESTAMP
-			);
-			INSERT INTO migrations (id, name, start_time, end_time) VALUES (0, 'init', CURRENT_TIMESTAMP, CURRENT_TIMESTAMP) ON CONFLICT DO NOTHING;
-			`)
-			return err
+			return nil
 		},
 	}
 }
 
 // ApplyMigrations does the necessary work to actualize state.
 // It will manipulate the objects available in baseEnv, and use the migrations table in db.
 func ApplyMigrations(ctx context.Context, db *sqlx.DB, baseEnv Env, state State) error {
+	tx, err := db.BeginTxx(ctx, &sql.TxOptions{})
+	if err != nil {
+		return err
+	}
+
+	env := baseEnv
+	env.Tx = tx
+
+	if _, err := tx.ExecContext(ctx, `CREATE TABLE IF NOT EXISTS migrations (
+				id BIGINT PRIMARY KEY,
+				NAME VARCHAR(250) NOT NULL,
+				start_time TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+				end_time TIMESTAMP
+			);`); err != nil {
+		return err
+	}
+
+	logrus.Infof("waiting to acquire lock on migrations table")
+	if _, err := tx.ExecContext(ctx, `LOCK TABLE migrations IN EXCLUSIVE MODE`); err != nil {
+		return err
+	}
+	logrus.Infof("acquired lock on migrations table")
+
 	for _, state := range collectStates(make([]State, 0, state.n+1), state) {
-		if err := applyMigration(ctx, db, baseEnv, state); err != nil {
-			return err
+		if err := applyMigration(ctx, tx, env, state); err != nil {
+			logrus.Error(err)
+			return tx.Rollback()
```

Comment:
I like doing the rollback here, it does seem cleaner.

Reply:
Although looking at it again, we were returning the proximate error before, and now we are returning the error from rollback.

I could be wrong here, but my understanding is that this would return nil for a successful rollback?  I think we want to return the error that caused the rollback, and log whether there was a problem rolling back.